src/routes.rs

```diff
@@ -1922,3 +2057,338 @@ pub(crate) async fn unlock(
+
+        let expiry = get_current_timestamp() + payload.timeout_secs as u64;
+        let swapstring = format!(
+            "{}/{}/{}/{}/{}/{}",
+            amt_asset,
+            asset_id,
+            swaptype.side(),
+            price,
+            expiry,
+            payment_hash.0.to_hex(),
+        );
+
+        let payment_secret = payment_secret.0.to_hex();
+        let payment_hash = payment_hash.0.to_hex();
+        Ok(Json(MakerInitResponse {
+            payment_hash,
+            payment_secret,
+            swapstring,
+        }))
+    })
+    .await
+}
+
+pub(crate) async fn taker(
+    State(state): State<Arc<AppState>>,
+    WithRejection(Json(payload), _): WithRejection<Json<TakerRequest>, APIError>,
+) -> Result<Json<TakerResponse>, APIError> {
+    use crate::swap::*;
+
+    no_cancel(async move {
+        let unlocked_state = state.check_unlocked().await?.clone().unwrap();
+        let ldk_data_dir_path = PathBuf::from(state.static_state.ldk_data_dir.clone());
+        let swapstring = SwapString::from_str(&payload.swapstring)
+            .map_err(|_| APIError::InvalidSwapString(payload.swapstring.clone()))?;
+
+        if get_current_timestamp() > swapstring.expiry {
+            return Err(APIError::Expired);
+        }
+
+        // We are selling assets, do we have enough?
+        if !swapstring.swap_type.is_buy() {
+            let max_balance = get_max_balance(
+                swapstring.asset_id,
+                &ldk_data_dir_path,
+                unlocked_state.channel_manager.list_channels().iter(),
+            );
+            if swapstring.swap_type.amount_rgb() > max_balance {
+                return Err(APIError::InsufficientAssets(
+                    max_balance,
+                    swapstring.swap_type.amount_rgb(),
+                ));
+            }
+        }
+
+        unlocked_state.taker_trades.lock().unwrap().insert(
+            swapstring.payment_hash,
+            (swapstring.asset_id, swapstring.swap_type),
+        );
+        let our_pubkey = unlocked_state.channel_manager.get_our_node_id();
+
+        Ok(Json(TakerResponse {
+            our_pubkey: our_pubkey.to_string(),
+        }))
+    })
+    .await
+}
+
+fn get_route(
+    channel_manager: &crate::ldk::ChannelManager,
+    router: &crate::ldk::Router,
+    start: PublicKey,
+    dest: PublicKey,
+    final_value_msat: Option<u64>,
+    asset_id: Option<ContractId>,
+    hints: Vec<RouteHint>,
+) -> Option<Route> {
+    let inflight_htlcs = channel_manager.compute_inflight_htlcs();
+    let payment_params = PaymentParameters {
+        payee: Payee::Clear {
+            node_id: dest,
+            route_hints: hints,
+            features: None,
+            final_cltv_expiry_delta: 14,
+        },
+        expiry_time: None,
+        max_total_cltv_expiry_delta: DEFAULT_MAX_TOTAL_CLTV_EXPIRY_DELTA,
+        max_path_count: 1,
+        max_channel_saturation_power_of_half: 2,
```

Comment:
These are both explain much better than I ever could on the ldk docs.

[`max_path_count`](https://docs.rs/lightning/latest/lightning/routing/router/struct.PaymentParameters.html#structfield.max_path_count) is the maximum number of MPP paths that can be used. Since we don't support MPP this is hardcoded to 1.
[`max_channel_saturation_power_of_half`](https://docs.rs/lightning/latest/lightning/routing/router/struct.PaymentParameters.html#structfield.max_channel_saturation_power_of_half) is how much of a channel's liquidity we can use in a single payment. 2 is the default value, so I used that. There's no right or wrong value here, if you make it too small you may try to use large chunks of liquidity in a channel, which will increase your failure rates because channels generally don't have 100% liquidity on one side. If you choose a value that's too high you may fail to route larger payments, because you will never try to use much of the liquidity available in the channels.

Reply:
thanks for explaining. it's a shame that the struct doesn't implement the Default trait and that the `DEFAULT_MAX_CHANNEL_SATURATION_POW_HALF` constant is not public, but not your fault